pennylane/tape/tapes/tape.py

```diff
@@ -96,6 +97,20 @@ def expand_tape(tape, depth=1, stop_at=None, expand_measurements=False):
 
     new_tape = tape.__class__()
 
+    if expand_measurements:
+        obs = [m.obs for m in tape.measurements if m.obs is not None]
+        m_idx = [i for i, m in enumerate(tape.measurements) if m.obs is not None]
+        obs_wires = [wire for m in tape.measurements for wire in m.wires if m.obs is not None]
+
+        if len(obs_wires) != len(set(obs_wires)):
+            rotations, diag_obs = diagonalize_qwc_pauli_words(obs)
+
+            tape._ops.extend(rotations)
+
+            for o, i in zip(diag_obs, m_idx):
+                new_m = qml.tape.measure.MeasurementProcess(tape.measurements[i].return_type, obs=o)
+                tape._measurements[i] = new_m
+
```

Comment:
Could this be made more efficient? It will be called at each depth level of expand?

Reply:
Good idea

Reply:
How about putting it in the `expand()` method, before `expand_tape()` is called?

Reply:
Worth trying, I think!